base/src/user_model/common.rs

```diff
@@ -882,6 +878,122 @@ impl UserModel {
                 });
             }
         }
+        // bottom of the cells above the first
```

Comment:
These sections seem to be repeating a good deal of shape/logic. Is it possible to pull out a helper method for setting these borders?